### ISSUE TITLE
Refresh access and id tokens in a React/Angular SPA

### ISSUE THREAD

**Issue by ravenscar** [aws/amazon-cognito-auth-js#92]
### Summary

I would say that without Cognito implementing `prompt=none` on the `/oauth2/authorize` endpoint, and whilst the `cognito` cookie on `<your_domain>.auth.<region>.amazoncognito.com/` expires after 60 minutes instead of 30 days (or what is set for the user pool) it is unsuitable to be used as an out-of-the-box solution for Single Page Applications.

### Details

The recommended OAuth2 flow for a Single Page Application (SPA) is the Implicit Grant. This flow, by design, does not issue a refresh token to the web app.

There are many issues in this project around refreshing tokens that seem to be resolved by switching to Authorization Code Grant instead, which is OK for some native apps and some non-SPA web apps (e.g. where there are round trips to a server) but is unsuitable for SPA apps.

It is considered unsafe and bad practice for a web application to store a refresh token in session storage, local storage, or in javascript accessible memory as this opens up the app to many attacks which could result in the refresh token being exposed to a malicious third party which is now able to obtain ID/API tokens for the user.

The OpenID Provider (OP) usually sets an http only secure cookie on user login, and this is used for refreshing in this case the OP is awscognito and the cookie would be set on `<your_domain>.auth.<region>.amazoncognito.com/`. Cognito does set this cookie.

The typical process for refreshing tokens in an SPA is to call the OP's `authorize` endpoint with a `prompt=none` in the query string. This would normally occur in an embedded iframe in the SPA so as to not alter the UI and would either return new tokens which can be stored or an error, which deletes the tokens.

This is not a standard or requirement as it is up the the OpenID Provider to determine how a user is authorized, however if this is not implemented and it means that an SPA must choose to log in whenever the access token expires, or to adopt poor practices and store refresh tokens in the SPA (there is another option where we could have our own server which stores the refresh tokens and can refresh api/id tokens and deliver them to the SPA).

This partially works using cognito hosted UI, there is a cognito cookie stored when you login at `<your_domain>auth.<region>.amazoncognito.com/login?client_id=<your_client_id>&response_type=token&redirect_uri=<your_redirect>` and if you subsequently call `<your_domain>auth.<region>.amazoncognito.com/oauth2/authorize?client_id=<your_client_id>&response_type=token&redirect_uri=<your_redirect>` it will redirect with new tokens, with updated timestamps, without requiring interaction.

Unfortunately the cognito token expires after 60 minutes so it can only be used to extend the session to a maximum of 120 minutes. Also since the `prompt` parameter is not implemented there is no way for the iframe to get an error so it cannot be used in an iframe.

### References

- [prompt=none behaviour, OpenId spec section 3.1.2.1.  Authentication Request](http://openid.net/specs/openid-connect-core-1_0.html)
- [AWS specs for authorization endpoint](https://docs.aws.amazon.com/cognito/latest/developerguide/authorization-endpoint.html) This does not detail prompt as an option and it appears to not be implemented.
- [Auth0 silent authentication for SPAs](https://auth0.com/docs/api-auth/tutorials/silent-authentication)
- [Okta silent authentication for SPAs](https://developer.okta.com/authentication-guide/tokens/refreshing-tokens#get-a-refresh-token-silently-for-your-spa)

**Comment by hhp21** [comment]
Any thoughts on this? It seems to me that it makes implicit grant totally impractical to use.

**Comment by salmonz** [comment]
I'd prefer if AWS allows the configuration the id_token and session cookie expiry TTL based on your own requirements.

**Comment by cervantek** [comment]
This seems like a deal breaker for using Cognito User Pools for SPA development.

**Comment by ppasmanik** [comment]
Why there is no response from aws team on this??? This is making implicit flow absolutely useless. I am not sure how @cervantek got cognito cookie to work for 8 hours.  I see only 1 hour, so you are stuck with maximum 2 hour for tokens if you do a workaround and make a call to oauth2/authorize endpoint for hosted login.

**Comment by luomavaltteri** [comment]
Wondering the same thing! This is a MUST feature for any real-world use.

**Comment by gercorr-gw** [comment]
This seems to be a vital feature for user experience. @cervantek do you have details on how to increase the cognito cookie timeout? But even that does not solve the issue.

**Comment by nihakue** [comment]
The document that you referenced assumes that refresh tokens last indefinitely, but I believe that with Cognito User Pools you can configure the refresh token TTL. https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-with-identity-providers.html#amazon-cognito-user-pools-using-the-refresh-token

**Comment by salmonz** [comment]
AWS needs to implement this properly by giving us the ability to configure cognito cookie timeout. If implemented, then using a proper Oidc JS library would get a new id_token and token periodically.

**Comment by Beretta1979** [comment]
We need this feature too

**Comment by ppasmanik** [comment]
I agree with @nihakue .  After talking to AWS team we went with that solution  - setting refresh token ttl to 1 day, and doing refresh token every hour with auth code flow.

**Comment by rajwilkhu** [comment]
Could someone from the AWS team respond to this? Came across this issue whilst trying to implement silent refresh. I am now considering moving away from Cognito to something that supports the implicit flow better for SPAs.

**Comment by maziarz** [comment]
While this may sound (very) crazy. What if the refreshToken had a more granular TTL, such as minutes or hours. Wouldn't this get us slightly closer to a silent auth/refresh solution?

**Comment by salmonz** [comment]
@maziarz while it might help, it's actually the wrong solution. Refresh tokens shouldn't be used in SPA apps; rather, use the session cookie controlling the refresh, e.g. cognito to use session cookie to see if session cookie is still valid and reissue another token and update session cookie if warranted.

**Comment by maziarz** [comment]
@salmonz its not that i disagree, i ran into this problem 1.5 years ago and ended up implementing Cognito with passport.js in the back utilising secure cookies. But eventually it removed all benefits from being truly "serverless" and having low maintenance on a SPA. I do not believe that anyone has solved this issue in an elegant way, even auth0 is relying on a valid SSO session in order for this to actually work with `prompt=none`. If we assume a session cookie and refreshToken has the same TTL, what would the difference be if we set the storage method to the side? Eventually the refreshToken becomes the cookie, don't you agree?

**Comment by maziarz** [comment]
@salmonz  its not that i disagree, i ran into this problem 1.5 years ago and ended up implementing Cognito with passport.js in the back utilising secure cookies. But eventually it removed all benefits from being truly "serverless" and having low maintenance on a SPA. I do not believe that anyone has solved this issue in an elegant way, even auth0 is relying on a valid SSO session in order for this to actually work with prompt=none. It does not look like Cognito team are going to introduce a cookie mechanism any time soon..

**Comment by salmonz** [comment]
AWS - do you have an update on this issue? Seems to me that the sign-in cookie expiry should be configurable in the AWS portal for a cognito instance, or per client configuration.

**Comment by docmoxie** [comment]
Any update on this?

**Comment by carvercj** [comment]
Any updates??

**Comment by vanpra1** [comment]
Any progress on this?

**Comment by valecarlos** [comment]
any updates?

**Comment by AshUK** [comment]
+1 for silent renew support on implicit grants

**Comment by dpistole** [comment]
Sub'd, this is a deal breaker for using cognito user pools with SPA, would be nice to hear something from AWS.

**Comment by sjbthfc2** [comment]
AWS: please give us some sort of update on this issue? What workarounds can you propose for SPA apps, do you have any plans to deal with this? The silence is deafening!

**Comment by jeremiahsmall** [comment]
I guess we'll have to use Auth0 again.

**Comment by sebastienfi** [comment]
> The typical process for refreshing tokens in an SPA is to call the OP's authorize endpoint with a prompt=none in the query string. **This would normally occur in an embedded iframe in the SPA** so as to not alter the UI and would either return new tokens which can be stored or an error, which deletes the tokens.

The _iFrame flow_ does not work on the latest Safari. 
Auth0 warns about it here: https://auth0.com/docs/api-auth/token-renewal-in-safari

So I guess all we have left is
> there is another option where we could have **our own server** which stores the refresh tokens and can refresh api/id tokens and deliver them to the SPA

**Comment by sebastienfi** [comment]
> I agree with @nihakue . After talking to AWS team we went with that solution - setting refresh token ttl to 1 day, and doing refresh token every hour with **auth code flow**.

@ppasmanik Can you please elaborate?

**Comment by ravenscar** [comment]
@sebastienfi 

I believe they are using the `Authorization Code Grant` instead of the `Implicit Grant` to get a code that can be exchanged for a refresh token, storing the refresh token in the SPA, and refreshing the access/id tokens hourly.

Storing the refresh token client side in a web app would normally be considered very insecure however they have limited the lifetime of the refresh token to 24 hours, which is a tradeoff they have had to make as AWS provides no silent auth.

I guess the upshot is that users would have to authenticate daily rather than hourly.

**Comment by sebastienfi** [comment]
@ravenscar Thank you for mentioning me - and for these clarifications!

**Comment by ppasmanik** [comment]
@ravenscar<https://github.com/ravenscar> You are correct. The cognito-auth-js library provides various APIs to get refresh token, session, etc.
It stores them in local storage. We wrote a small library that wraps amazon-cognito-auth-js and provides React components that know how to handle both types of the flows depending on configuration, perform refresh of tokens using oauth2/authorize endpoint of hosted cognito at configurable intervals (between 10 and 55 minutes depending on user roles).
The minimum refresh token ttl is 1 day for cognito pool and it is sufficient for our users.  You can set it to longer if you need. We also integrated idle timeout into that library that logs a user out after a configurable time interval.
________________________________

**Comment by jeremiahsmall** [comment]
Storing secrets in local storage is the entire problem. The workarounds described are too insecure for some to consider.

**Comment by ppasmanik** [comment]
Yes, storing secrets in local storage is not a good practice, however, it is questionable whether refresh token with validity limited to a set number of hours is really a secret.
You need both unexpired token and refresh token to renew a token.
If someone is able to get hold of an unexpired token, he will be able to get in.
If someone is able to get hold of both an unexpired token and refresh token, he will be able to refresh the token several times up to a refresh token expiration time.
In both cases, something or somebody needs to hack something to get hold of tokens from local storage.

________________________________

**Comment by leotohill** [comment]
> 
> 
> Any thoughts on this? It seems to me that it makes implicit grant totally impractical to use.

Actually, I think you mean that the auth code grant is impractical.  Only the auth code grant provides a refresh token.  Also, the implicit grant is inadvisable these days.

**Comment by leotohill** [comment]
@ppasmanik , "You need both unexpired token and refresh token to renew a token."
Actually, that's not correct.  You need only the refresh token. See section 6 of https://tools.ietf.org/html/rfc6749,  also https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/

**Comment by ravenscar** [comment]
> Actually, I think you mean that the auth code grant is impractical. Only the auth code grant provides a refresh token. Also, the implicit grant is inadvisable these days.

@leotohill not sure I follow you here. Auth code grant cannot be used securely in a SPA _because_ it returns a refresh token which cannot be securely stored.

When implemented correctly by an auth provider implicit grant can allow refresh without the token being stored in a place accessible to javascript, this is desirable if you are not auditing every node module in your dependency tree.

Do you have a recent source saying why this is suddenly inadvisable? What is the alternative "safer" way of storing refresh tokens in an app where all javascript code cannot be trusted.

**Comment by leotohill** [comment]
@ravenscar yes, [this ietf best practice document.](https://datatracker.ietf.org/doc/draft-ietf-oauth-browser-based-apps/?include_text=1).  The auth code grant is a more secure flow, and the refresh token can (should be) discarded, or not issued in the first place if that's an option. 

Not sure i fully understand what you are saying about the auth provider "allow refresh without...", but could this work anyway, even if the token is acquired via the auth code flow?

I too find that this advice makes things challenging.  I'm thinking of setting up an server where the refresh token can be stored server-side, and the server executes the refresh operation, returning the new token to the client.  This would require a cookie-based session with the server.

**Comment by ravenscar** [comment]
@leotohill interesting read but not sure it's worth trading implicit grant headaches for CORS headaches. Maybe in 5 years if CORS allows multiple origins.

I am not sure if the silent refresh would work with auth-code flow. I can't see a reason it wouldn't, but I guess since it's optional it's up to the implementer.

Your idea seems to be re-implementing the out-of-the-box functionality of other auth providers such as Auth0. I'd advise just to use their services and forget that cognito user pools exists.

Even when you get past this hurdle you have a dozen more, such as discovering that the sign-up and log-in emails are case sensitive, so `joe@gmail.com` can't login if he signed up as `Joe@gmail.com`.

Nothing AWS has produced has disappointed me as much as cognito user pools. Shouldn't event be in beta.

**Comment by ajhool** [comment]
I submitted a ticket to AWS support center, which hopefully has a channel to a different Cognito team than this open-source sdk team.

AWS makes a(n arbitrary IMO) distinction between open-sourced and closed-source/backend services, so I urge everybody else to open support tickets. Particularly if you work at a large enough company that you have the top-tier of customer service or work at a mid-large company and depend on cognito.

I agree with the other sentiment that cognito is a fundamental service, a fundamentally flawed service, and the one that makes me wonder if AWS has stretched itself too thin but the alternative providers in this area all have issues of their own, too.

**Comment by evbo** [comment]
fyi, AWS' official client library for cognito was able to confirm that better support for SPA authentication use-cases is in the Cognito team's backlog:
https://github.com/aws-amplify/amplify-js/issues/3436#issuecomment-538400098
